internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,674 @@
+	"github.com/openfga/openfga/internal/concurrency"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+)
+
+const IteratorMinBatchThreshold = 1000
+
+type fastPathSetHandler func(ctx context.Context, iterQueue []*iteratorProducer, iterChan chan *iteratorMsg)
+
+type iteratorMsg struct {
+	iter storage.TupleKeyIterator
+	err  error
+}
+
+type iteratorProducer struct {
+	idx      int
+	iter     storage.TupleKeyIterator
+	done     bool
+	producer chan *iteratorMsg
+}
+
+// NOTE: caller should consider running this in a goroutine to not block.
+func cleanupIteratorProducers(iterProducers []*iteratorProducer) {
+	for _, iterProducer := range iterProducers {
+		if iterProducer.iter != nil {
+			iterProducer.iter.Stop()
+		}
+		for msg := range iterProducer.producer {
+			if msg.iter != nil {
+				msg.iter.Stop()
+			}
+		}
+	}
+}
+
+// pollIteratorProducers will return a list of the remaining active producers.
+// To be considered active your producer channel must still be open.
+func pollIteratorProducers(ctx context.Context, iterProducers []*iteratorProducer) ([]*iteratorProducer, error) {
+	for _, producer := range iterProducers {
+		if producer.iter != nil || producer.done {
+			// no need to poll further
+			continue
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case i, ok := <-producer.producer:
+			if !ok {
+				producer.done = true
+				break
+			}
+			if i.err != nil {
+				return nil, i.err
+			}
+			producer.iter = i.iter
+		}
+	}
+	// TODO: in go1.23 compare performance vs slices.Collect
+	// clean up all empty entries that are both done and drained
+	return slices.DeleteFunc(iterProducers, func(entry *iteratorProducer) bool {
+		return entry.done && entry.iter == nil
+	}), nil
+}
+
+func (c *LocalChecker) fastPathDirect(ctx context.Context,
+	req *ResolveCheckRequest) (chan *iteratorMsg, error) {
+	typesys, _ := typesystem.TypesystemFromContext(ctx)
+	ds, _ := storage.RelationshipTupleReaderFromContext(ctx)
+	tk := req.GetTupleKey()
+	objRel := tuple.ToObjectRelationString(tuple.GetType(tk.GetObject()), tk.GetRelation())
+	i, err := checkutil.IteratorReadStartingFromUser(ctx, typesys, ds, req, objRel, nil)
+	if err != nil {
+		return nil, err
+	}
+	iterChan := make(chan *iteratorMsg, 1)
+	if !concurrency.TrySendThroughChannel(ctx, &iteratorMsg{iter: i}, iterChan) {
+		i.Stop() // will not be received to be cleaned up
+	}
+	close(iterChan)
+	return iterChan, nil
+}
+
+func (c *LocalChecker) fastPathComputed(ctx context.Context, req *ResolveCheckRequest, rewrite *openfgav1.Userset) (chan *iteratorMsg, error) {
+	typesys, _ := typesystem.TypesystemFromContext(ctx)
+	tk := req.GetTupleKey()
+	computedRelation := rewrite.GetComputedUserset().GetRelation()
+	rewrittenTupleKey := tuple.NewTupleKey(
+		tk.GetObject(),
+		computedRelation,
+		tk.GetUser(),
+	)
+
+	childRequest := req.clone()
+	childRequest.TupleKey = rewrittenTupleKey
```

Comment:
Nit: we have fixed (https://github.com/openfga/openfga/issues/1995) the data race that appeared on cloning a `req` and then trying to modify the tuple. You can just do:

```
childRequest := req.clone()
childRequest.TupleKey.Relation = computedRelation
```